packages/x-tree-view/src/internals/plugins/useTreeViewKeyboardNavigation/useTreeViewKeyboardNavigation.ts

```diff
@@ -51,16 +50,13 @@ export const useTreeViewKeyboardNavigation: TreeViewPlugin<
 
     const newFirstCharMap: { [itemId: string]: string } = {};
 
-    const processItem = (item: TreeViewBaseItem) => {
-      const getItemId = params.getItemId;
-      const itemId = getItemId ? getItemId(item) : (item as { id: string }).id;
-      newFirstCharMap[itemId] = instance.getNode(itemId).label!.substring(0, 1).toLowerCase();
-      item.children?.forEach(processItem);
+    const processItem = (node: TreeViewNode) => {
+      newFirstCharMap[node.id] = node.label!.substring(0, 1).toLowerCase();
     };
 
-    params.items.forEach(processItem);
+    Object.values(state.nodes.nodeMap).forEach(processItem);
```

Comment:
Btw this will support stuff like `apiRef.current.updateItem` when we add this kind of API, the previous approach did not.

It's still far from perfect because we loop through the whole dataset, but until we actually have partial dataset update it's good enough.